### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by safafa**
In this project, we implemented our own version of the enumerable methods.

### CODE REVIEW

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle:

**Good effort so far** :+1:

**It's impressive that your methods are exhibiting some of the behaviors as expected.**

**Please see below required changes to make your enumerables cover all test cases and behave exactly like in built ruby enumerables.**

Firstly,
- [x] The workflow linters file is currently wrongly named. It's "linters.yml" and not "linterns.yml". Please rename it.

Below are method-specific required changes. Your methods should do the following in addition to what they are currently doing.

**my_each**
- [x] my_each should return an Enumerator if no block is given.
- [x] When a block is given, when self is an array, it should return the array itself after calling the given block once for each element in self.
- [x] my_each when a block is given, when self is a range, it should return the range itself after calling the given block once for each element in self.
- [x] my_each when a block is given, when self is a range, it behaves like #each, it should call the given block once for each element in self.
- [x] when a block is given, when self is a hash, it behaves like #each, it should call the given block once for each element in self.

**my_each_with_index**

- [x] my_each_with_index should return an enumerator if no block is given.
- [ ] When a block is given, when self is an array, it should return the array itself after calling the given block once for each element in self.
- [ ] When a block is given, when self is an array, it should behave like #each_with_index, it should call the given block once for each element in self.
- [ ]  When a block is given, when self is a range, it should return the range itself after calling the given block once for each element in self.
- [ ] When a block is given, when self is a range, it behaves like #each_with_index, it should call the given block once for each element in self
- [ ] When a block is given, when self is a hash, it behaves like #each_with_index, it should call the given block once for each element in self.

**my-select**

- [ ] my_select should return an array containing all elements of enum for which the given block returns a true value::range.
- [ ]  It should return an enumerator if no block is given.

**my_all?**

- [ ] It should return false if the block returns false or nil::range.
- [ ] When a class is passed as an argument, it should return true if all of the collection is a member of such class::Integer
- [ ]  When a class is passed as an argument, it should return true if all of the collection is a member of such class::Numeric
- [ ] When a class is passed as an argument, it should return false if any of the collection is not a member of such class.
- [ ] When a Regex is passed as an argument, it should return true if all of the collection matches the Regex.
- [ ]  When a Regex is passed as an argument, it should return false if any of the collection does not match the Regex.
- [ ]  When a pattern other than Regex or a Class is given, it should return true if all of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return false if any of the collection does not match the pattern.

**my_any?**

- [ ] It should return false if the block ever returns a value false or nil::range.
- [ ] When a class is passed as an argument, it should return true if at least one of the collection is a member of such class::Numeric.
- [ ] When a class is passed as an argument, it return true if at least one of the collection is a member of such class::Integer.
- [ ] When a Regex is passed as an argument, it should return true if any of the collection matches the Regex.
- [ ] When a Regex is passed as an argument, it should return false if none of the collection matches the Regex.
- [ ]  When a pattern other than Regex or a Class is given, it should return false if none of the collection matches the pattern.
- [ ]  When a pattern other than Regex or a Class is given, it should return true if any of the collection matches the pattern.

**my_none?**
- [ ] It should return false if the block ever returns true for all elements::range.
- [ ]  When a class is passed as an argument, it should return true if none of the collection is a member of such class.
- [ ] When a class is passed as an argument, it should return true if none of the collection is a member of such class::Numeric.
- [ ] When a class is passed as an argument, it should return false if any of the collection is a member of such class.
- [ ] When a Regex is passed as an argument, it should return true if none of the collection matches the Regex.
- [ ] When a Regex is passed as an argument, it should return false if any of the collection matches the Regex.
- [ ] When a pattern other than Regex or a Class is given, it should return true only if none of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return false only if one of the collection matches the pattern.

**my_map**

- [ ] my_map should return a new array with the results of running a given block once for every element in enum::range
- [ ] It should return an Enumerator if no block is given.
- [ ] Executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

**my_inject**

- [ ] When a block is given without an initial value, it combines all elements of enum by applying a binary operation, specified by a block::range
- [ ] When a block is given with an initial value, it combines all elements of enum by applying a binary operation, specified by a block.
- [ ]  When a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method.
- [ ] When a symbol is specified with an initial value, it combines each element of the collection by applying the symbol as a named method.
- [ ] When a symbol is specified with an initial value, it combines each element of the collection by applying the symbol as a named method::range.

Please make the required changes and submit another CRR when you're ready.

Please reach out to me on slack if you require further clarification. 

Happy coding. :desktop_computer:

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    end
+    self
+  end
+
+  def my_each_with_index
+    i = 0
+    return enum_for unless block_given?
+
+    arr = to_a if is_a?(Range) || is_a?(Hash)
+    while i < arr.size
```
Comment (Berabjesus) [inline main.rb:31]: - [x] I am getting `noMethodError` when i run this.
```
  arr = [1,2,3,4,5]
  arr.my_each_with_index {|num| num + 1}

```
and
```
array.my_each_with_index(&SomeTestblock)

```
Please test it and fix the bug.

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      my_each { |item| return false if item.class != args }
+    elsif !args.nil? && instance_of?(Hash)
+      return false unless empty?
+    elsif !args.nil? && args.instance_of?(Regexp)
+      my_each { |item| return false unless args.match(item.to_s) }
+    else
+      my_each { |item| return false if item != args }
+    end
+    true
+  end
```
Comment (Berabjesus) [inline main.rb]: - [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric.
This test should return true.
```
  arr = [1,2,3,4,5]
  p arr.my_all?(Numeric)

```

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      elsif !args.nil? && args.instance_of?(Regexp)
+        my_each { |item| return true if args.match(item.to_s) }
+      elsif args.nil?
+        my_each { |item| return true if item == args }
+      else
+        my_each { |item| return true if item }
+      end
+    end
+    false
+  end
```
Comment (Berabjesus) [inline main.rb]: - [x] This test should return false

```
p [nil, false, nil, false].my_any?

```

- [x] I am getting a false result here, but there is at least one numeric value.

```
  arr = [1,2,3,4,5]
  p arr.my_any?(Numeric)
```

- [x] The test below should return false. check with the original any?

```
p   %w[dog door rod blade].my_any?('cat')

```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      elsif !sym.nil? && sym.instance_of?(Symbol)
+        my_each { |item| result = item.send(sym, result) }
+      end
+      return result
+    end
+    return to_a if is_a?(Hash)
+
+    my_each { |item| result = yield(result, item) }
+    result
+  end
```
Comment (Berabjesus) [inline main.rb]: - [x] my_inject searches for the longest word in an array of strings.
```
  search = proc { |memo, word| memo.length > word.length ? memo : word }
  p %w[dog door rod blade].my_inject(&search)
```
The above code should return `blade`. Check with the original inject.

- [x] my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation,
```
  range = Range.new(5, 50) 

 actual = range.my_inject { |prod, n| prod * n }
 expected = range.inject { |prod, n| prod * n }
```
expected result: `1267253883404724085150525340252698701849068398706688000000000000`
gotten `0`

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: CHANGES REQUIRED :warning: :construction: 
### Hi @safafa and @srizzuto great job on working on the changes required earlier. 

### However, out of 79 tests 12 are failing so you do have a few changes to carry out before we proceed so please take a look at the things you have to work on and their respective review comments below. Check if your enum methods satisfy the following requirements.

- [x] my_each_with_index should not return `NoMethodError: undefined method size for nil:NilClass` on three occasions (check explanation below). 
- Because of the above issue the following are also failing: -

1. my_each_with_index does not mutate the original array.
2. my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self.
3. my_each_with_index when a block is given when self is an array behaves like each_with_index calls the given block once for each element in self

- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of a given class. (check comment below)
- [x] my_any? when no block or argument is given returns false if at least one of the collection is not true. (check comment below)
- [x] my_any? when a class is passed as an argument returns true if at least one of the collections is a member of a given class.  (check comment below)
- [x] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [ ] my_inject raises a "LocalJumpError" when no block or argument is given.
- [x] my_inject searches for the longest word in an array of strings. (check comment below)
- [x] my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation,  (check comment below)
- [x] my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol 
- [x] multiply_els accepts an array as an argument and multiplies all the elements of the array together using my_inject

#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Comment by srizzuto** [comment]
All changes made as required

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
# Status: Requested Changes 

Good job @safafa and @srizzuto your methods are working and you worked on the requested changes by TSEs but there is a small change you need to make and your project will be approved.

## Kindly fix that and submit another CRR

Slack: Delice Lydia
[Linkedin](https://www.linkedin.com/in/delice-lydia/)
[Twitter](https://twitter.com/IngabireLydia3)

**Comment by DeliceLydia** [comment]
inject

my_inject method needs to raise an error when no block is given

**Comment by safafa** [comment]
Fixed as required @DeliceLydia Thanks for the review

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
+# rubocop: disable Metrics/ModuleLength
```
Comment (Dipeshtwis) [inline main.rb:1]: [Optional]
You can fix that by using best practices in ruby.

**Review by Dipeshtwis — APPROVED** [review]
# Status: Approved ✔️

## Hi @safafa &  @srizzuto

Good job on implementing the suggested changes. 👏👏👏👏

## [Optional]

- [ ]  RuboCop cops are meant to catch bad code and enforce best practices and the rubocop.yml file provided by [Microverse linters repo](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) doesn't enforce strict rules.
By disabling RuboCop rules, the student is suppressing the potential warnings they’d be getting, missing out on making their Ruby code more idiomatic.
(See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no-ignore-linters.png) from [pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/))
_TBA_


**You may close this Pull Request and merge the changes**
  
**Keep being awesome devs and good luck with your next milestone!**
---------

Cheers and Happy coding!👏👏👏

Code review by **Dipesh Kumar**
(Feel free to contact me on slack by the handle of  **deep** if you have any additional questions).
[Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)

**Pull request merged**